### PR TITLE
fix: variable_validation_for_service_endpoints

### PR DESCRIPTION
**Pull request by Khuzaima05**
### Description

This module has a service_endpoints variable that allows values of public, private and public-and-private. Data engine does not currently support private endpoints, so this var should be updated to only allow public

issue: https://github.com/terraform-ibm-modules/terraform-ibm-data-engine/issues/47
#### Changes that affect the core Terraform module or submodules
- [x] Bug fix
- [ ] New feature
- [ ] Dependency update

#### Changes that don't affect the core Terraform module or submodules
- [ ] Examples or tests (addition or updates of examples or tests)
- [ ] Documentation update
- [ ] CI-related update (pipeline, etc.)
- [ ] Other

#### Release required?
Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning).

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

If a release is required, replace this text with information that users need to know about the release. Write the release notes to help users understand the changes, and include information about how to update from the previous version.

Your notes help the merger write the commit message for the PR that is published in the release notes for the module.

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### Merge actions for mergers

- Use a relevant [conventional commit](https://www.conventionalcommits.org/) message that is based on the PR contents and any release notes provided by the PR author. The commit message determines whether a new version of the module is needed, and if so, which semver increment to use (major, minor, or patch).
- Merge by using "Squash and merge".

### CODE REVIEW

**Review by ocofaigh — CHANGES_REQUESTED** [review]
@Khuzaima05 I just realised that the `service_endpoints` variable is not actually used for the data engine instance. Its used to determine what KMS service endpoint to use when using KMS encryption. So I suggest what we do is rename the variable to `kms_endpoint` and only allow `public` and `private` values.
Propose this for new variable description:
`The KMS endpoint to use when configuring KMS encryption. Must be private or public.`

**Review by ocofaigh — CHANGES_REQUESTED** [review]
@Khuzaima05 There is another bug that has just been reported. This is wrong:
`kms_rootkey_id : var.kms_key_crn`

The value should be in ID, not the CRN. They key will create with CRN, but when you look at the instance it can’t access the key. Can you please update the variable to `kms_key_id` and verify it works please?

**Comment by Khuzaima05** [comment]
> [terraform-ibm-data-engine] Fix the variable validation for service_endpoints #47



> @Khuzaima05 There is another bug that has just been reported. This is wrong: `kms_rootkey_id : var.kms_key_crn`
> 
> The value should be in ID, not the CRN. They key will create with CRN, but when you look at the instance it can’t access the key. Can you please update the variable to `kms_key_id` and verify it works please?

ok will check.

File: variables.tf [hunk 1]
```diff
@@ -81,3 +81,9 @@ variable "kms_key_crn" {
     error_message = "Value must be the root key CRN from Key Protect"
   }
 }
+
```
Comment (ocofaigh) [inline variables.tf:84]: You need to remove `kms_key_crn` now - it has been replaced by `kms_key_id`. You will also need to update the validation to now use the `kms_key_id` variable

File: variables.tf [hunk 2]
```diff
@@ -81,3 +81,9 @@ variable "kms_key_crn" {
...
+variable "kms_key_id" {
+  type        = string
+  description = "The id of the root key in the KMS instance that will be used to encrypt the cluster."
```
Comment (ocofaigh) [inline variables.tf:87]: cluster?? You mean the data engine. Lets go with:

The root key ID of a Key Protect key that you want to use to encrypt your stored Data Engine jobs. Only used if var.kms_encryption_enabled is set to true. NOTE: Hyper Protect Crypto Services is not currently supported by Data Engine.

**Review by ocofaigh — CHANGES_REQUESTED** [review]
@Khuzaima05 see latest comments. Also can you confirm you tested using the key ID and it worked? Can you show proof (screenshot) that the data engine instance is indeed encrypted with the key protect key? Thanks

**Comment by ocofaigh** [comment]
We should also do some investigation into what impact this update would have on an existing instance:

```
Messages:   	Resource(s) identified to be updated Name: data_engine_instance Address: module.data_engine.ibm_resource_instance.data_engine_instance Actions: [update]
        	            	DIFF:
        	            	Before: {"parameters":{"customerKeyEncrypted":"true","kms_instance_id":"{\"guid\":\"a19c425c-2cdb-4a0b-815d-faf26bd0ab34\",\"url\":\"[https://private.eu-de.kms.cloud.ibm.com\](https://private.eu-de.kms.cloud.ibm.com/)"}","kms_rootkey_id":"crn:v1:bluemix:public:kms:eu-de:a/abac0df06b644a9cabc6e44f55b3880e:a19c425c-2cdb-4a0b-815d-faf26bd0ab34:key:5fe6bf6b-152a-4de8-8355-11f460c7af24"}}
        	            	After: {"parameters":{"customerKeyEncrypted":"true","kms_instance_id":"{\"guid\":\"a19c425c-2cdb-4a0b-815d-faf26bd0ab34\",\"url\":\"[https://private.eu-de.kms.cloud.ibm.com\](https://private.eu-de.kms.cloud.ibm.com/)"}","kms_rootkey_id":"5fe6bf6b-152a-4de8-8355-11f460c7af24"}}
```

Is it going to cause downtime to the existing instance? We will need to document the impact in the release notes

**Comment by Khuzaima05** [comment]
> @Khuzaima05 see latest comments. Also can you confirm you tested using the key ID and it worked? Can you show proof (screenshot) that the data engine instance is indeed encrypted with the key protect key? Thanks

Screenshot 2023-08-17 at 12 01 27 PM

**Comment by ocofaigh** [comment]
Thanks for the screenshot @Khuzaima05 - please re-request a review once all review comments are addressed

**Comment by ocofaigh** [comment]
I'm doing a manual upgrade test to understand what the consumers experience is going to be when upgrading in order to be able to draft the release notes

**Comment by dpittner** [comment]
Created related bug in data engine, we should have never accepted the old request and deny creation: https://github.ibm.com/SqlServiceWdp/main/issues/7859 also the instances created that way will be unusable, so re-creating them should not cause issues for folks.

**Comment by ocofaigh** [comment]
Skipping the upgrade test since the release notes are going to instruct consumers they need to destroy any old instance which attempted to do KMS encryption in order to consume the new version with the fix in it. Probably not a big deal since the instance was not usable anyway

**Review by ocofaigh — APPROVED** [review]

**Pull request merged**